README.md
Neutron-Capture Nucleosynthesis and the Chemical Evolution of Globular Clusters
==================================
Australian National University, 2015, PhD Thesis of Luke Jeremy Shingles

Published at http://hdl.handle.net/1885/16507


Abstract
---------
Elements heavier than iron are almost entirely produced in stars through neutron captures and radioactive decays. Of these heavy elements, roughly half are produced by the slow neutron-capture process (s-process), which takes place under extended exposure to low neutron densities. Most of the s-process production occurs in stars with initial masses between roughly 0.8 and 8 solar masses (Msun), which evolve through the Asymptotic Giant Branch (AGB) phase.

This thesis explores several topics related to AGB stars and the s-process, with a focus on comparing theoretical models to observations in the literature on planetary nebulae, post-AGB stars, and globular cluster stars. A recurring theme is the uncertainty of carbon-13-pocket formation, which is crucial for building accurate models of s-process nucleosynthesis.

We first investigated whether neutron-capture reactions in AGB stars are the cause of the low sulphur abundances in planetary nebulae and post-AGB stars relative to the interstellar medium. Accounting for uncertainties in the size of the partial mixing zone that forms carbon-13 pockets and the rates of neutron-capture and neutron-producing reactions, our models failed to reproduce the observed levels of sulphur destruction. From this, we concluded that AGB nucleosynthesis is not the cause of the sulphur anomaly. We also discovered a new method to constrain the extent of the partial mixing zone using neon abundances in planetary nebulae.

We next aimed to discover the stellar sites of the s-process enrichment in globular clusters that have inter- and intra-cluster variation, with the examples of M4 (relative to M5) and M22, respectively. Using a new chemical evolution code developed by the candidate, we tested models with stellar yields from rotating massive stars and AGB stars. We compared our model predictions for the production of s-process elements with abundances from s-poor and s-rich populations. We found that rotating massive stars alone do not explain the pattern of abundance variations in either cluster, and that a contribution from AGB stars with carbon-13 pockets is required. We derived a minimum enrichment timescale from our best-fitting chemical evolution models and, although the value depends on the assumptions made about the formation of carbon-13 pockets, our estimate of 240 to 360 Myr for M22 is consistent with the upper limit of 300 Myr inferred by isochrone fitting.

Lastly, there is accumulating evidence that some stars (e.g., in Omega Centauri) have been born with helium mass fractions as high as 40%. This motivated us to explore the impact of helium-rich abundances on the evolution and nucleosynthesis of intermediate-mass (3 to 6 Msun) AGB models. We found that the stellar yields of s-process elements are substantially lower in He-rich models, largely as a result of less intershell material being mixed into the envelope. We also found evidence that high He abundances could restrict the s-process production by carbon-13 pockets to stars with lower initial masses.


changes.txt
Changes from the initial submitted copy to the final library copy:

Reviewer 1:
- Fixed equation 1.6 (was actually the inverse)
- Page 8, beginning of Sect 1.4.3. Revised wording.

Reviewer 3:
- Changed Figure 1.2 to Timmes other version showing s-process path (to match caption)
- Fixed Figure 1.4 caption (top -> left, bottom -> right)
- Added vertical axis label to Figure 7

Other changes
- Modified acknowledgments
- Fixed arrow type in Figure 1.5 for Rb86 to Sr86


generateplots/plotcommon.py
#!/usr/bin/env python3

import struct

dashesList = [(),(1.5,2,9,2),(5,1),(0.5,2),(4,2)]
dash_capstyleList = ['butt','butt','butt','round','butt']
colorList = ['black',(0,.8*158./255,0.6*115./255),(204./255,121./255,167./255),(213./255,94./255,0.0)]
markersList = ['s','D','o','^','v','*']
linewidth = 1.3
markersize = 3

fruitylabel = 'SCP14'
fruitylinecolor = '0.20' #or 'blue'
fruitydashes = (1,1)
fruitymarker = 'None'
fruitylinewidth = linewidth

VD09label = 'VD09'
VD09linecolor = '0.30' #or 'red'
VD09dashes = (3,1)
VD09marker = 'None'
VD09linewidth = linewidth

#surface abundances colour
abundancecolorlist = ['black',(0.0,0.4,0.6),(0.35,0.7,0.9),(0.9,0.2,0.0),(0.9,0.6,0.0),(0.0,0.6,0.5), (0.6,0.2,0.7)] #(0.95,0.9,0.25)]
abundancedasheslist = [()]*7

#surface abundances B&W version
#abundancecolorlist = ['black']*4 + ['0.4']*3
#abundancedasheslist = [(),(6,1,0.7,1),(6,1),(3,1)]*2

fs = 10 # font size
fsticklabel = 8

framewidth = 1

columnwidth = 3 # in inches for figsize=(w,h)
textwidth = 6.2

singlefigsize = (columnwidth,columnwidth*0.9)
doublefigsize = (columnwidth,columnwidth*1.7)
triplefigsize = (columnwidth,columnwidth*2)     #mgisotopes, #core mass luminosity
#quadfigsize = (columnwidth,columnwidth*2.5) #core mass luminosity

doublewidefigsize = (textwidth,columnwidth)

def getVariable(filecontent, byteoffset, variablecode):
    startpos = byteoffset + 4 + (variablecode * 4)
    return struct.unpack('<f', filecontent[startpos:startpos+4])[0]


generateplots/plottpcycle.py
#!/usr/bin/env python3
import struct
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
from plotcommon import *

everynmodels = 1

timetp1 = []
mcoretp1 = []

evrootfolder = '/Users/lukes/Dropbox/Papers (first author)/2015 He-enhanced IMAGB Stars/generateplots/evolution/'

modelname = 'm3z0006y24'

# get time at TP1
with open(evrootfolder + modelname + "/iscvn.dat", mode='r') as file:
    for i in range(6):
        row = file.readline().split()
    row = file.readline().split()
    timetp1 = float(row[1])+34
    mintime = timetp1 - 50
    timeselectedtp = float(file.readline().split()[1])
    maxtime = timeselectedtp + 150

#listtimetp1[1] -= 2400

# read ev file
listTime = []
listceinner = []
listmhcore = []
listmhecore = []
listconvtime = [[]]
listconvinner = [[]]
listconvouter = [[]]
listL = []
listLH = []
listLHe = []
listLC = []
with open(evrootfolder + modelname + "/evall.dat", mode='rb') as file:
    print('Processing ' + modelname + "...")
    fileContent = file.read()
    print('Starting mass: %.3f' % struct.unpack("d", fileContent[4:12])[0])

    #b is the byte number for the current model
    for b in range(20,len(fileContent)-16,268*everynmodels): #268 means every model
#            nmod = struct.unpack("i", fileContent[b:b+4])[0]
        T = struct.unpack("d", fileContent[b+4:b+12])[0]
        Toffset = (T-timetp1) / 1.0e0 # scaled relative time

        if T <= maxtime:
            if T >= mintime:
                convinner = getVariable(fileContent,b,9)
                convouter = getVariable(fileContent,b,8)
                if (convouter - convinner) > 1e-5: # convection zone exists below the envelope
                    # end of one convection zone and start of another
                    if len(listconvinner[-1]) > 0 and ((convouter < listconvinner[-1][-1] or convinner > listconvouter[-1][-1])):
                        # when line plotting, close up top and bottom of convective zones at start and end
#                        listconvtime[-1].append(listconvtime[-1][-1])
#                        listconvinner[-1].append(listconvinner[-1][-1])
#                        listconvouter[-1].append(listconvinner[-1][-1])
#                        listconvtime.append([listTime[-1]])
#                        listconvinner.append([convinner])
#                        listconvouter.append([convinner])
                        listconvtime.append([])
                        listconvinner.append([])
                        listconvouter.append([])
                    listconvtime[-1].append(Toffset)
                    listconvinner[-1].append(convinner)
                    listconvouter[-1].append(convouter)
                elif len(listconvinner[-1]) > 0:
                    #signal end the current conv. zone by making a new empty one
                    listconvtime.append([])
                    listconvinner.append([])
                    listconvouter.append([])

                mtot = getVariable(fileContent,b,60)
#                listTime.append(T)
                listTime.append(Toffset)
                listceinner.append(getVariable(fileContent,b,7))
                listmhcore.append(getVariable(fileContent,b,5))
                listmhecore.append(getVariable(fileContent,b,10))
                listL.append(getVariable(fileContent,b,15))
                listLH.append(getVariable(fileContent,b,16))
                listLHe.append(getVariable(fileContent,b,17))
                listLC.append(getVariable(fileContent,b,18))
                
                # keep the base of the convective envelope fixed if it disappears
                if len(listceinner) > 1 and (mtot - listceinner[-1]) < 1e-2:
                    listceinner[-1] = listceinner[-2]
        else:
            break

assert len(listconvtime) == len(listconvinner) == len(listconvouter)
print("%d time samples" % len(listTime))

listtimeranges = [[listTime[0],listTime[980]],[listTime[980],listTime[-990]],[listTime[-990],listTime[-1]]]

fig, axes = plt.subplots(2, len(listtimeranges), figsize=(5,4), sharex=False, sharey=False, tight_layout={"pad":0.6, "w_pad":0.0, "h_pad":0.0})

modellabel = modelname[1] + ' M$_\odot$ $Y=0.' + modelname[-2:] + '$'

for nx in range(len(axes[0])): #column within the row
    for ny in range(len(axes)): #row
        plt.setp(axes[ny][nx].get_xticklabels(), fontsize=fsticklabel)
        plt.setp(axes[ny][nx].get_yticklabels(), fontsize=fsticklabel)
    #    ax.xaxis.set_minor_locator(ticker.MultipleLocator(base=2.0))
        axes[ny][nx].xaxis.set_major_locator(ticker.MultipleLocator(base=[25,10000,50][nx]))
        axes[ny][nx].set_xlim(xmin=listtimeranges[nx][0],xmax=listtimeranges[nx][1])
        axes[ny][nx].get_yaxis().set_label_coords(-0.12,0.5)
        axes[ny][nx].get_xaxis().get_major_formatter().set_useOffset(False)
        if nx != 0:
            plt.setp(axes[ny][nx].get_yticklabels(), visible=False)
        if ny != len(axes)-1:
            plt.setp(axes[ny][nx].get_xticklabels(), visible=False)
        for axis in ['top','bottom','left','right']:
            axes[ny][nx].spines[axis].set_linewidth(framewidth)

    axes[0][nx].set_yscale('linear')
    axes[0][nx].yaxis.set_minor_locator(ticker.MultipleLocator(base=0.1))
    axes[0][nx].set_ylim(ymin=min(listmhecore)*0.999,ymax=max(listceinner)*1.002)

    axes[0][nx].plot(listTime, listmhcore,  color='red', lw=1.0, label='H-exhausted core')
    axes[0][nx].plot(listTime, listmhecore, color='blue', lw=1.0, label='He-exhausted core')
    axes[0][nx].fill_between(listTime, listceinner, [max(listceinner)*2 for x in listceinner], facecolor=(0.55,0.80,0.64), interpolate=True, edgecolor='None')
    axes[0][nx].plot(listTime, listceinner, color=(0.09,0.5,0.16), markersize=markersize*0.5, marker='o', markeredgewidth=0, linestyle='None')

    for n in range(len(listconvinner)-1):
        axes[0][nx].fill_between(listconvtime[n], listconvinner[n], listconvouter[n], facecolor=(0.48,0.73,0.54), interpolate=True, edgecolor='None')
        axes[0][nx].plot(listconvtime[n], listconvinner[n], color=(0.09,0.5,0.16), markersize=markersize*0.5, marker='o', markeredgewidth=0, linestyle='None')
        axes[0][nx].plot(listconvtime[n], listconvouter[n], color=(0.09,0.5,0.16), markersize=markersize*0.5, marker='o', markeredgewidth=0, linestyle='None')

    axes[1][nx].set_yscale('log')
    axes[1][nx].set_ylim(ymin=1e0,ymax=max(listLHe)*1.1)
    labellist = ['$\mathrm{L}_\mathrm{H}$','$\mathrm{L}_\mathrm{He}$']
    for (yList,color,dashes,label) in zip([listLH,listLHe],['red','blue'],[(),(4,1)],labellist):
        axes[1][nx].plot(listTime, yList, color=color, marker='None', lw=linewidth, dashes=dashes, label=label)

axes[-1][1].set_xlabel(r'Time [yr]', fontsize=fs)
axes[-1][1].set_xticks(axes[-1][1].get_xticks()[1:-1])

axes[0][1].legend(loc='upper left',handlelength=1,frameon=False,numpoints=1,prop={'size':fs-2}, markerscale=3)
axes[0][0].set_ylabel(r'$\mathrm{M}_\mathrm{r} \,[\mathrm{M}_\odot]$', fontsize=fs)
axes[1][0].set_ylabel(r'$\mathrm{L} \,[\mathrm{L}_\odot]$', fontsize=fs)
axes[0][0].get_yaxis().set_label_coords(-0.26,0.5)
axes[1][0].get_yaxis().set_label_coords(-0.26,0.5)
axes[1][1].legend(loc='upper left',handlelength=1.5,frameon=False,numpoints=1,prop={'size':fs-1})

    #plt.locator_params(axis = 'y', nbins = 30)

fig.savefig('../chapter1/fig-tpcycle.pdf',format='pdf')
plt.close()


generateplots/plotsolarabund.py
#!/usr/bin/env python3
import math
import struct
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
from plotcommon import *

elements = ('','H','He','Li','Be','B','C','N','O','F','Ne','Na','Mg','Al','Si','P',
        'S','Cl','Ar','K','Ca','Sc','Ti','V','Cr','Mn','Fe','Co','Ni','Cu',
        'Zn','Ga','Ge','As','Se','Br','Kr','Rb','Sr','Y','Zr','Nb','Mo','Tc','Ru',
        'Rh','Pd','Ag','Cd','In','Sn','Sb','Te','I','Xe','Cs','Ba','La',
        'Ce','Pr','Nd','Pm','Sm','Eu','Gd','Tb','Dy','Ho','Er','Tm','Yb','Lu',
        'Hf','Ta','W','Re','Os','Ir','Pt','Au','Hg','Tl','Pb','Bi','Po','At',
        'Rn','Fr','Ra','Ac','Th','Pa','U')

listatomicnumber = []
listabundance = []

highlightedelements = ['H','C','O','Fe','Sr','Ba','Pb','Th']
highlightedatomicnumbers = [elements.index(x) for x in highlightedelements]
highlightedelementyposition = [0.0 for x in highlightedelements]

fig, ax = plt.subplots(1, 1, sharey=True, figsize=(5,3), tight_layout={"pad":0.2,"w_pad":0.0,"h_pad":0.0})

with open('solar_abund_ARAAedit.dat', mode='r') as file:
    for line in file.readlines():
        row = line.split()
        if len(row) > 1:
            abundance = float(row[3])
            atomicnumber = int(row[0])
#            if abundance > -5:
            listatomicnumber.append(atomicnumber)
            listabundance.append(abundance)

            for highlightedatomicnumber in highlightedatomicnumbers:
                if abs(atomicnumber-highlightedatomicnumber) <= 1: #is the element or a close neighbour
                    index = highlightedatomicnumbers.index(highlightedatomicnumber)
                    if abundance > highlightedelementyposition[index]:
                        highlightedelementyposition[index] = abundance

tcindex = listatomicnumber.index(43) #Tc
pmindex = listatomicnumber.index(61) #Pm
# skip Tc and Pm
for ((startindex,stopindex)) in [[0,tcindex],[tcindex+1,pmindex],[pmindex+1,len(listatomicnumber)]]:
    ax.plot(listatomicnumber[startindex:stopindex], listabundance[startindex:stopindex], marker='o', markersize=markersize*1.5, markeredgewidth=0, color='blue', lw=linewidth)

plt.setp(plt.getp(ax, 'xticklabels'), fontsize=fsticklabel)
plt.setp(plt.getp(ax, 'yticklabels'), fontsize=fsticklabel)
for axis in ['top','bottom','left','right']:
    ax.spines[axis].set_linewidth(framewidth)
#    ax.annotate(modellabel, xy=(0.05, 0.93), xycoords='axes fraction', horizontalalignment='left', verticalalignment='top', fontsize=fs)
    #ax.set_yscale('log')

for (x,y,symbol) in zip(highlightedatomicnumbers,highlightedelementyposition,highlightedelements):
    ax.annotate(symbol, xy=(x, y - 0.0 * (x % 2)), xycoords='data', textcoords='offset points', xytext=(0,10), horizontalalignment='center', verticalalignment='center', weight='bold', fontsize=fs-1.5)

ax.set_xlabel("Atomic number", fontsize=fs)
ax.xaxis.set_minor_locator(ticker.MultipleLocator(base=5))
ax.set_xlim(xmin=0,xmax=95)

ax.set_ylabel(r'log(X/H) + 12', fontsize=fs)
ax.yaxis.set_minor_locator(ticker.MultipleLocator(base=0.5))
ax.set_ylim(ymin=-0.5,ymax=13.5)

fig.savefig('../chapter1/fig-solarabundances.pdf',format='pdf')
plt.close()

appendices/makeyieldtextables.py
#!/usr/bin/env python3
from collections import OrderedDict
import os
import struct

yieldfilelist = []
for (dirpath, dirnames, filenames) in os.walk("shinglesetal2015onlinedata/"):
    yieldfilelist.extend(filter(lambda fn: fn[-4:]=='.txt', filenames))
    break

with open('yieldtables.tex', 'w') as fileout:
    fileout.write('\n')

    prevmass = '-1'
    for yieldfilename in yieldfilelist:
        mass = yieldfilename[10:].split('z')[0]
#        if mass != prevmass:
#            fileout.write(r'\section{' + mass + ' \Msun models}' + '\n')
        prevmass = mass
    
        modelproperties = []
        with open(dirpath + yieldfilename,'r') as fyields:
            print("reading " + yieldfilename)
            for i in range(3):
                modelproperties.append(fyields.readline().split())
            fyields.readline()
            
            #tablecaption = yieldfilename
            tablecaption = '{Yields for ' + mass + r' \Msun, $Z=0.0006$, $Y=' + modelproperties[0][10].rstrip(',') + '$'
            if mass == '3':
                mpmz = float(modelproperties[0][13])
                if mpmz > 1e-6:
                    tablecaption += ', $\M{pmz}=' + '{0:.3f}'.format(mpmz) + '$'
                else:
                    tablecaption += ', no PMZ.'
            tablecaption += '}'
#            fileout.write(r'\begin{longtable}{r r r r r r r}' + '\n')
#            fileout.write(r'\caption{' + tablecaption + '.}' + '\n')
            fileout.write(r'\clearpage\ctable[cap = ' + tablecaption + ', caption = ' + tablecaption + ', label = {tab:' + yieldfilename[:-4] + '}]{r r r r r r r}{}{' + '\n')
            fileout.write(r'\hline' + '\n')
            fileout.write(r'El&   $Z$&    log $\epsilon$(X)&[X/H]&   [X/Fe]&M$_\mathrm{yield}$&X$_\mathrm{yield}$\\' + '\n')
            fileout.write(r'\hline' + '\n')
            
            for line in fyields:
                if not line.startswith("#") and not line.startswith('   '): #and not line.startswith(' bi  83'):
                    row = line.split()
                    row[0] = row[0].capitalize()
                    if row[1] == '1':
                        row[0] = 'H'
                    fileout.write("&".join(row) + r'\\' + '\n')

                    if row[1] == '42':
                        fileout.write(r'\hline' + '\n')
                        fileout.write(r'}' + '\n')
                        fileout.write(r'\ctable[cap = , caption = ' + tablecaption + ', label = {tab:' + yieldfilename[:-4] + '-cont}, continued]{r r r r r r r}{}{' + '\n')
                        fileout.write(r'\hline' + '\n')
                        fileout.write(r'El&   $Z$&    log $\epsilon$(X)&[X/H]&   [X/Fe]&M$_\mathrm{yield}$&X$_\mathrm{yield}$\\' + '\n')
                        fileout.write(r'\hline' + '\n')

            fileout.write(r'\hline' + '\n')
            fileout.write(r'}' + '\n\n')
#            fileout.write(r'\end{longtable}' + '\n')


generateplots/plotagbfinalabund.py
#!/usr/bin/env python3
import math
import struct
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
from plotcommon import *

elements = ('','H','He','Li','Be','B','C','N','O','F','Ne','Na','Mg','Al','Si','P',
        'S','Cl','Ar','K','Ca','Sc','Ti','V','Cr','Mn','Fe','Co','Ni','Cu',
        'Zn','Ga','Ge','As','Se','Br','Kr','Rb','Sr','Y','Zr','Nb','Mo','Tc','Ru',
        'Rh','Pd','Ag','Cd','In','Sn','Sb','Te','I','Xe','Cs','Ba','La',
        'Ce','Pr','Nd','Pm','Sm','Eu','Gd','Tb','Dy','Ho','Er','Tm','Yb','Lu',
        'Hf','Ta','W','Re','Os','Ir','Pt','Au','Hg','Tl','Pb','Bi','Po','At','Rn','Fr','Ra','Ac')

highlightedelements = ['Fe','Rb','Y','Ba','Ce','Pb'] #'He','N','F',
highlightedatomicnumbers = [elements.index(x) for x in highlightedelements]
highlightedelementyposition = [0.0 for x in highlightedelements]

fig, ax = plt.subplots(1, 1, sharey=True, figsize=(5,3), tight_layout={"pad":0.2,"w_pad":0.0,"h_pad":0.0})

dppnspath = '/Users/lukes/Dropbox/Papers (first author)/2015 He-enhanced IMAGB Stars/generateplots/ppns320species/'

modelnames = ['m1.7z0006y24a4pmz001s320','m6z0006y24a0s320']
modellabels = ['1.7 M$_\odot$, Z=0.0006','6.0 M$_\odot$, Z=0.0006']
for (modelname,modellabel,color) in zip(modelnames,modellabels,['red','blue']):
    listatomicnumber = []
    listabundance = []
    with open(dppnspath + modelname + '/final.dat', mode='r') as file:
        for line in file.readlines():
            row = line.split()
            if len(row) > 1 and not line.startswith('#'):
                abundance = float(row[4])
                atomicnumber = int(row[1])
                if abundance > -5 and atomicnumber != 84:
                    listatomicnumber.append(atomicnumber)
                    listabundance.append(abundance)
                    for highlightedatomicnumber in highlightedatomicnumbers:
                        if abs(atomicnumber-highlightedatomicnumber) <= 1: #is the element or a close neighbour
                            index = highlightedatomicnumbers.index(highlightedatomicnumber)
                            if abundance > highlightedelementyposition[index]:
                                highlightedelementyposition[index] = abundance
    
    tcindex = listatomicnumber.index(elements.index('Tc'))
    pmindex = listatomicnumber.index(elements.index('Pm'))
    # skip Tc and Pm
    for ((startindex,stopindex)) in [[0,tcindex],[tcindex+1,pmindex],[pmindex+1,len(listatomicnumber)]]:
        ax.plot(listatomicnumber[startindex:stopindex], listabundance[startindex:stopindex], marker='o', markersize=markersize*1.5, markeredgewidth=0, color=color, lw=linewidth, label=(modellabel if startindex==0 else ''))

plt.setp(plt.getp(ax, 'xticklabels'), fontsize=fsticklabel)
plt.setp(plt.getp(ax, 'yticklabels'), fontsize=fsticklabel)
for axis in ['top','bottom','left','right']:
    ax.spines[axis].set_linewidth(framewidth)

for (x,y,symbol) in zip(highlightedatomicnumbers,highlightedelementyposition,highlightedelements):
    ax.annotate(symbol, xy=(x, y - 0.0 * (x % 2)), xycoords='data', textcoords='offset points', xytext=(0,10), horizontalalignment='center', verticalalignment='center', weight='bold', fontsize=fs-1.5)

#ax.set_yscale('log')
ax.set_xlabel('Atomic number', fontsize=fs)
ax.xaxis.set_minor_locator(ticker.MultipleLocator(base=5))
ax.set_xlim(xmin=25,xmax=85)

ax.set_ylabel(r'[X/Fe]', fontsize=fs)
ax.yaxis.set_minor_locator(ticker.MultipleLocator(base=0.5))
ax.set_ylim(ymin=-0.1,ymax=3.4)
ax.legend(loc='upper left',handlelength=2,frameon=False,numpoints=1,prop={'size':fs-1})

fig.savefig('../chapter1/fig-agbfinalabund.pdf',format='pdf')
plt.close()

generateplots/plotc13pocket.py
#!/usr/bin/env python
import csv
import sys
import math
import struct
import numpy as np
import matplotlib.pyplot as plt
from os import walk
from plotcommon import *

def isBinaryFile (filename):
    fin = open(filename, 'rb')
    try:
        CHUNKSIZE = 1024
        while 1:
            chunk = fin.read(CHUNKSIZE)
#            if '\0' in chunk: # found null byte
            if b'\x00' in chunk: # found null byte
                return True
            if len(chunk) < CHUNKSIZE:
                break # done
    finally:
        fin.close()
    return False

#nmodstart = 304320
#nmodend   = 306000
nmodstart = -1
nmodend = 999999999

speciesNamesDisplayed = ('p','c13','n14')
colorList = ['black','red','blue']

#modelfolder = "/Volumes/Seagate 2TB/coala/z0006models-herich/320species/m6z0006y35a0s320/"
modelfolder = "m3z0006y24a0pmz001s320/"

speciesfile = modelfolder + "species.dat"

with open(speciesfile, mode='r') as speciesFile:
#    csvReader = csv.reader(speciesFile, delimiter=' ', skipinitialspace=True)
#    speciesCount = int(csvReader.next()[0])
    line = speciesFile.readline()
    speciesCount = int(line.strip())
    speciesList = [()] * speciesCount
    print("# species file: '" + speciesfile + "' lists " + str(speciesCount) + " species")
    for i in range(speciesCount):
        #mass number, elm symbol, species name, neutron number
#        row = csvReader.next()
        row = speciesFile.readline().split()
        speciesList[i] = (int(row[0]), row[1], row[2], int(row[3]))

#convert species names to numbers
speciesDisplayedNumbers = list(map(lambda z: speciesList.index(z), filter(lambda z: z[2] in speciesNamesDisplayed, speciesList)))

#compfilelist = []
#for (dirpath, dirnames, filenames) in walk(modelfolder + "comp/"):
#    compfilelist.extend(filter(lambda fn: fn[:2]=='ns' and fn[-4:]=='.cmp' and int(fn[-11:-4]) >= nmodstart and int(fn[-11:-4]) <= nmodend, filenames))
#    break

fig, axes = plt.subplots(1, 2, sharey=True, figsize=(6,3), tight_layout={"pad":0.1,"w_pad":0.3,"h_pad":0.3})

dirpath = modelfolder
compfilelist = ["ns320_0003550.cmp","ns320_0003677.cmp"]
for (ax, compfile) in zip(axes,compfilelist):
    if isBinaryFile(dirpath + compfile):
        print("\n==> comp file (bin): '" + compfile)
        with open(dirpath + compfile, mode='rb') as file:
            fileContent = file.read()
            #model number is wrong for some reason
            modelNumber = struct.unpack("<H", fileContent[0:2])[0]
            numMassPoints = struct.unpack("<H", fileContent[2:4])[0]
            compNumSpecies = struct.unpack("<H", fileContent[4:6])[0]
            massPoints = struct.unpack("f" * numMassPoints, fileContent[6:6+4*numMassPoints])
            abundances = np.zeros((numMassPoints,len(speciesDisplayedNumbers)))
            
            for massPointNum in range(numMassPoints):
                for s in range(len(speciesDisplayedNumbers)):
                    #one is up flows, other is down flows
                    abundIndex = 6 + 4 * (numMassPoints + compNumSpecies * massPointNum + speciesDisplayedNumbers[s])
                    abundIndex2 = 6 + 4 * (numMassPoints + compNumSpecies * (numMassPoints + massPointNum) + speciesDisplayedNumbers[s])
                    massnumber = speciesList[speciesDisplayedNumbers[s]][0]
                    #average up and down flows and multiply by mass number to get mass fraction
                    abundances[massPointNum][s] = (massnumber * 0.5 * (struct.unpack("<f", fileContent[abundIndex:abundIndex+4])[0] + struct.unpack("f", fileContent[abundIndex2:abundIndex2+4])[0]))
            
            convindex = 6 + 4 * (numMassPoints + numMassPoints*compNumSpecies*2)
            numConvectiveBoundaries = struct.unpack("H", fileContent[convindex:convindex+2])[0]
            convectiveBoundaries = struct.unpack("f" * numConvectiveBoundaries, fileContent[convindex + 2:convindex + 2 + 4*numConvectiveBoundaries])
            convectiveBoundaries = zip(convectiveBoundaries[0::2], convectiveBoundaries[1::2])
    else:
        print("\n==> comp file (text): " + compfile)
        with open(dirpath + compfile, 'r') as txtFile:
            csvReader = csv.reader(txtFile, delimiter=' ', skipinitialspace=True)
            
            compRaw = []
            for row in csvReader:
                for col in row:
                    if col != '':
                        compRaw.append(col)
        modelNumber = int(compRaw[0])
        numMassPoints = int(compRaw[1])
        compNumSpecies = int(compRaw[2])
        massPoints = map(float, compRaw[3:3+int(compRaw[1])])
        abundances = np.zeros((numMassPoints,len(speciesDisplayedNumbers)))

        for massPointNum in range(numMassPoints):
            for s in range(len(speciesDisplayedNumbers)):
                massnumber = speciesList[speciesDisplayedNumbers[s]][0]
                abundances[massPointNum][s] = (massnumber * 0.5 * (float(compRaw[3 + numMassPoints + massPointNum*compNumSpecies + speciesDisplayedNumbers[s]]) + float(compRaw[3 + numMassPoints + (numMassPoints + massPointNum)*compNumSpecies + speciesDisplayedNumbers[s]])))
        numConvectiveBoundaries = int(compRaw[3 + numMassPoints + numMassPoints*compNumSpecies*2])
        convectiveBoundaries = list(map(float,compRaw[3 + numMassPoints + numMassPoints*compNumSpecies*2+1:][:numConvectiveBoundaries]))
        convectiveBoundaries = zip(convectiveBoundaries[0::2], convectiveBoundaries[1::2])

    modelNumber = int(compfile[-11:-4]) #overwrite value from the file with number from filename

    print("# model number:", modelNumber)
    print("# number of mass points:", numMassPoints)
    print("# comp file species count:", compNumSpecies)
    print("# convective boundaries: [" + "][".join(['%.5f, %.5f' % (cvz[0],cvz[1]) for cvz in convectiveBoundaries]) + "]")
    #print "# values are mass fraction"
    #print ",".join(["#mass"]+map(lambda x:speciesList[x][2],speciesDisplayedNumbers))

    for cvz in convectiveBoundaries:
        #plt.bar(cvz[0], 1, width=cvz[1]-cvz[0], bottom=10**-24, color='1.0',lw=1,alpha=1, hatch="//", edgecolor='0.4')
        ax.bar(cvz[0], 1, width=cvz[1]-cvz[0], bottom=10**-24,lw=0, alpha=1.0, color=(0.55,0.80,0.64))

    for s in range(len(speciesDisplayedNumbers)):
        linelabel = speciesNamesDisplayed[s]
        if linelabel == 'p':
            linelabel = 'H'
        else:
            linelabel = linelabel.title()
        ax.plot(massPoints, [massShell[s] for massShell in abundances], color=colorList[s], label=linelabel, lw=linewidth)

    ax.set_yscale('log')
    ax.legend(loc='upper left',ncol=1,handlelength=1.5,frameon=False,numpoints=1,prop={'size':fs-1})
    ax.set_xlim(xmin=0.7865,xmax=0.789)
    ax.set_ylim(ymin=1.0e-13,ymax=1.0)
    plt.setp(plt.getp(ax, 'xticklabels'), fontsize=fs)
    plt.setp(plt.getp(ax, 'yticklabels'), fontsize=fs)
    ax.set_xticks(ax.get_xticks()[1:-2])
#    ax.set_yticks(ax.get_yticks()[2:-2])
    for axis in ['top','bottom','left','right']:
        ax.spines[axis].set_linewidth(framewidth)
    ax.set_xlabel(r'$\mathrm{M}_\mathrm{r} \,[\mathrm{M}_\odot]$', fontsize=fs)
axes[0].set_ylabel('Mass fraction', fontsize=fs)

fig.savefig('../chapter1/fig-c13pocket.pdf', format='pdf')
plt.close()